src/modules/job-info/job-info.c

```diff
@@ -0,0 +1,411 @@
+static void lookups_cancel (struct info_ctx *ctx,
+                            const char *sender, uint32_t matchtag)
+{
+    struct lookup_ctx *l;
+
+    l = zlist_first (ctx->lookups);
+    while (l) {
+        lookup_cancel (ctx, l, sender, matchtag);
+        l = zlist_next (ctx->lookups);
+    }
+}
+
+static void cancel_cb (flux_t *h, flux_msg_handler_t *mh,
+                       const flux_msg_t *msg, void *arg)
+{
+    struct info_ctx *ctx = arg;
+    uint32_t matchtag;
+    char *sender;
+
+    if (flux_request_unpack (msg, NULL, "{s:i}", "matchtag", &matchtag) < 0) {
+        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
+        return;
+    }
+    if (flux_msg_get_route_first (msg, &sender) < 0) {
+        flux_log_error (h, "%s: flux_msg_get_route_first", __FUNCTION__);
+        return;
+    }
+    lookups_cancel (ctx, sender, matchtag);
+    free (sender);
+}
+
+static void disconnect_cb (flux_t *h, flux_msg_handler_t *mh,
+                           const flux_msg_t *msg, void *arg)
+{
+    struct info_ctx *ctx = arg;
+    char *sender;
+
+    if (flux_request_decode (msg, NULL, NULL) < 0) {
+        flux_log_error (h, "%s: flux_request_decode", __FUNCTION__);
+        return;
+    }
+    if (flux_msg_get_route_first (msg, &sender) < 0) {
+        flux_log_error (h, "%s: flux_msg_get_route_first", __FUNCTION__);
+        return;
+    }
+    lookups_cancel (ctx, sender, FLUX_MATCHTAG_NONE);
+    free (sender);
+}
+
+static void stats_cb (flux_t *h, flux_msg_handler_t *mh,
+                      const flux_msg_t *msg, void *arg)
+{
+    struct info_ctx *ctx = arg;
+
+    if (flux_respond_pack (h, msg, "{s:i}",
+                           "lookups", zlist_size (ctx->lookups)) < 0) {
+        flux_log_error (h, "%s: flux_respond_pack", __FUNCTION__);
+        goto error;
+    }
+
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, NULL) < 0)
+        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
+}
+
+static const struct flux_msg_handler_spec htab[] = {
+    { .typemask     = FLUX_MSGTYPE_REQUEST,
+      .topic_glob   = "job-info.eventlog-lookup",
+      .cb           = lookup_cb,
+      .rolemask     = 0
+    },
+    { .typemask     = FLUX_MSGTYPE_REQUEST,
+      .topic_glob   = "job-info.eventlog-cancel",
+      .cb           = cancel_cb,
+      .rolemask     = 0
+    },
+    { .typemask     = FLUX_MSGTYPE_REQUEST,
+      .topic_glob   = "job-info.disconnect",
+      .cb           = disconnect_cb,
+      .rolemask     = 0
+    },
+    { .typemask     = FLUX_MSGTYPE_REQUEST,
+      .topic_glob   = "job-info.stats.get",
+      .cb           = stats_cb,
+      .rolemask     = 0
+    },
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+static void info_ctx_destroy (struct info_ctx *ctx)
+{
+    if (ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        if (ctx->lookups)
+            zlist_destroy (&ctx->lookups);
```

Comment:
Any in progress lookups should probably be canceled here (and removed from list and destroyed).

Reply:
I've set a free function on the zlist entries so they are freed via that mechanism.  But you're right, should cancel any watches.